### ISSUE TITLE
Base rule extension: [func-style] Allow type-annotated variable declaration (Follow-up)

### ISSUE THREAD

**Issue by NatoBoram** [typescript-eslint/typescript-eslint#10127]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/?=extension#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/func-style

### Description

If the option is set to `declaration`, an expression would be allowed if the variable has a type annotation.

### Fail

```TypeScript
const badFunc = function () {
	// …
}

const badLambda = () => {
	// …
}
```

### Pass

```TypeScript
function goodFunc() {
	// …
}

const goodTypedFunc: SomeType = function () {
	// …
}

const goodTypedLambda: SomeType = () => {
	// …
}
```

### Additional Info

* Follow-up on https://github.com/typescript-eslint/typescript-eslint/issues/8488#issuecomment-2016975772 because the conversation has been unnecessarily locked
* Upstream rejected the proposal in https://github.com/eslint/eslint/issues/18842#issuecomment-2400677082

cc @alfaproject, @anthonwellsjo, @auvred, @kirillgroshkov, @mrnagydavid, @nikitalpopov, @remcohaszing

**Comment by bradzacher** [comment]
We (the maintenance team) chatted about this internally.

To summarise - we don't think there's a strong enough usecase for us to action this within this project.

It sucks that upstream ESLint doesn't want to support this, but I understand they're also weary of their maintenance surface.

Thus we are going to reject this proposal for our project.
If this is important to you, we encourage you to build it using our tooling and publish it as a 3rd party plugin for others to use.

---

The maintenance burden of an extension rule is very high - several hundred lines of implementation and tests -- and that's assuming we can do an extension and don't have to fork the rule entirely to implement this feature. Extension rules can be high-maintenance too as we need to keep them synced with upstream changes and sometimes breakages occur (i.e. when they change selectors) and things are complicated to support them across the ESLint major versions we support.

This is also a purely stylistic lint rule. Generally speaking now-a-days we don't want more stylistic rules in our project for various reasons I won't get into here.

Extension rules are also high-burden for user onboarding as they need to discover them and override configs to use the extension rules. This can lead to low usage for extension rules with niche purposes as people don't know they could switch to them unless they run into the reason an extension was created.

I say "niche" above because as mentioned in the other thread -- this project has existed for >8y now, and this wasn't first requested till this year, and then in 5 months the issue only gathered 6 unique community interactions. Additionally based on the queries I made in the other thread -- there isn't a lot of usage of the rule. So there isn't a major community driver for this. 

So summarising -- the maintenance cost is high and the broad user value appears to be low. For this reason we shall reject this extension rule.
